app/controllers/concerns/v2/authentication.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module V2
+  # Authentication logic for all controllers
+  module Authentication
+    extend ActiveSupport::Concern
+
+    ALLOWED_CERT_BASED_RBAC_ACTIONS = [
+      { controller: 'policies', action: 'index' },
+      { controller: 'tailorings', action: 'tailoring_file' }
+    ].freeze
+
+    included do
+      around_action :authenticate_user
+    end
+
+    def authenticate_user
+      if identity_header.blank?
+        return unauthenticated(error: 'Error parsing the X-RH-IDENTITY header')
+      end
+      return unauthenticated unless identity_header.valid?
+      return unauthorized unless rbac_allowed?
+
+      return if performed?
+
+      set_authenticated_user
+      yield
+    ensure
+      User.current = nil
+    end
+
+    def current_user
+      User.current
+    end
+
+    def unauthorized(error: 'User is not authorized to view this page')
+      render(
+        json: { error: "Authorization error: #{error}" },
+        status: :forbidden
+      )
+      false
+    end
+
+    def unauthenticated(error: 'X-RH-IDENTITY header should be provided')
+      render(
+        json: { error: "Authentication error: #{error}" },
+        status: :unauthorized
+      )
+      false
+    end
+
+    def user
+      @user ||= User.new(account: Account.from_identity_header(identity_header))
+    end
+
+    private
+
+    def set_authenticated_user
+      User.current = user
+    end
+
+    def valid_cert_endpoint?
+      ALLOWED_CERT_BASED_RBAC_ACTIONS.include?(
+        controller: controller_name, action: action_name
+      )
+    end
+
+    def any_inventory_hosts?
+      Insights::Api::Common::HostInventory.new(b64_identity: raw_identity_header).hosts.dig('results').present?
```

Comment:
This is an external service that we cannot test, so please wrap `# :nocov:` around it